### DISCUSSION TITLE
Drift Postgres - What is it missing to be stable?

### DISCUSSION THREAD

**Discussion by ynnob** [simolus3/drift#2218] [Q&A] [answered]
Hey @simolus3 !
I am so glad to stumble upon the experimental feature for postgres integration by chance. I had almost given up. All other generators and semi orms were a disaster and since I've been using drift for sqlite for a long time I was really happy to see this!

So far i found no issues for basic cases when using `extras/drift_postgres`. What is it missing to be "stable"? 
Are you going to implement it into drift or is there no further development and it was just a test?

Thanks for you contribution to the dart/flutter community!

**Comment by simolus3** [comment, marked as answer]
Hi, thanks! To give credit where it's most due, @westito has done most of the work with the postgres implementation. Also, using it (and reporting issues) is a great help to get it stabilized.

There are some remaining issues with the postgres implementation that I know of:

1. Sometimes, types for variables are not set correctly: This is actually a limitation in the `postgres` package, which actually does its own in-place query manipulation in an attempt to figure out the right types. This is not needed with `drift` (since we _know_ the correct type to use). I wanted to fix this upstream in https://github.com/isoos/postgresql-dart/pull/73, but the consensus was that it's better to clean up the overall API in that package first, and that requires a large re-implementation in that package. I have started with some work on that, but I didn't have the time to finish it yet.
2. Drift's API is heavily tailored towards `sqlite3`. This is not _too_ much of an issue since sqlite3 in many cases stays close to Postgres, but some features like date manipulation are very different in sqlite3. We need to figure out a way to either transpile these functions to work on postgres at runtime, or find a better way to not expose them at all with postgres (without making drift harder to use with sqlite3).
3. The same is true for the generator: We only currently parse the sqlite3 SQL dialect. We also transform queries defined in `.drift` files, so we might even break correctly-defined queries on postgres.

So while there is nothing fundamentally wrong with the postgres version, I want the stable parts of drift to be pleasant and reliable. That is my goal for the postgres integration as well, but it's not quite there yet.

**Reply by westito** [reply]
1. Postgres query bind placeholders can have type annotations (`INSERT INTO table (id) VALUES (@a:int4)`). Does it help? 
2. Don't think we need to give a full-featured Postgres support. I suggest to add a "limited" label instead of experimental and clarify these limitations. One can use custom SQL queries for special cases that are not go through processing. 
3. Same as above

**Reply by simolus3** [reply]
I tried to use ` @1:bytea` but that blows up in `Query.validateParameters`: The postgres package doesn't send the type to the server in the `Bind` or `Prepare` message, but then expects that the types postgres infers (always `text` because it's not specified) matches the declares types...
